registry/storage/metadata/metadata.go

```diff
@@ -0,0 +1,59 @@
+package metadata
+
+import (
+	"fmt"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/reference"
+)
+
+type ErrKeyNotFound struct {
+	Key string
+}
+
+func (err ErrKeyNotFound) Error() string {
+	return fmt.Sprintf("metadata key not found: %s", err.Key)
+}
+
+// Metadatable signifies metadata support
+type Metadatable interface {
+	MetadataService() MetadataService
+}
+
+type MetadataUpdateRecord struct {
+	Actual   interface{}
+	Expected interface{}
+}
+
+// IterFunc is called by Iterate.  A non-nil error stops iteration
+type IterFunc func(string, interface{}) error
+
+// MetadataService defines the operations of a metadata service
+type MetadataService interface {
+	// Put puts a value under a key
+	Put(context.Context, string, interface{}) error
+
+	// BatchPut performs a transactional put of multiple key value pairs
+	BatchPut(context.Context, map[string]MetadataUpdateRecord) error
+
+	// Get gets a value for a given key
+	Get(context.Context, string) (interface{}, error)
```

Comment:
It may make sense to make `Get` return a deep copy of the object that's stored, so modifications to the result of `Get` don't corrupt the stored object. This is probably only an issue with the `inmemory` store.

It would mean operating on some interface type with `Copy` method instead of `interface{}`.

This is the approach we took in swarmkit, and even though it introduces a lot of overhead, it was probably the right choice. It's really easy to accidentally modify an object returned by `Get`.